sui/deploy-contract.js

```diff
@@ -23,49 +54,179 @@ async function processCommand(contractName, config, chain, options) {
+            const signers = await getSigners(keypair, config, chain, options);
+
+            validateParameters({ isNonEmptyString: { previousSigners, minimumRotationDelay }, isKeccak256Hash: { domainSeparator } });
+
+            const [creatorCap, relayerDiscovery, upgradeCap] = getObjectIdsByObjectTypes(publishTxn, [
+                `${packageId}::gateway::CreatorCap`,
+                `${packageId}::discovery::RelayerDiscovery`,
+                '0x2::package::UpgradeCap',
+            ]);
+
+            const encodedSigners = signersStruct
+                .serialize({
+                    ...signers,
+                    nonce: bytes32Struct.serialize(signers.nonce).toBytes(),
+                })
+                .toBytes();
+
+            const tx = new Transaction();
+
+            const separator = tx.moveCall({
+                target: `${packageId}::bytes32::new`,
+                arguments: [tx.pure(arrayify(domainSeparator))],
+            });
+
+            tx.moveCall({
+                target: `${packageId}::gateway::setup`,
+                arguments: [
+                    tx.object(creatorCap),
+                    tx.pure.address(operator),
+                    separator,
+                    tx.pure.u64(minimumRotationDelay),
+                    tx.pure.u64(options.previousSigners),
+                    tx.pure(bcs.vector(bcs.u8()).serialize(encodedSigners).toBytes()),
+                    tx.object('0x6'),
+                ],
+            });
+
+            if (policy !== '0') {
+                const upgradeType = policy === '128' ? 'only_additive_upgrades' : 'only_dep_upgrades';
+
+                tx.moveCall({
+                    target: `0x2::package::${upgradeType}`,
+                    arguments: [tx.object(upgradeCap)],
+                });
+            }
+
+            const result = await broadcast(client, keypair, tx);
+
+            printInfo('Setup transaction digest: ', result.digest);
+
+            const [gateway] = getObjectIdsByObjectTypes(result, [`${packageId}::gateway::Gateway`]);
+
+            contractConfig.objects = {
+                gateway,
+                relayerDiscovery,
+                upgradeCap,
+            };
+            contractConfig.domainSeparator = domainSeparator;
+            contractConfig.operator = operator;
+            contractConfig.minimumRotationDelay = minimumRotationDelay;
+            break;
+        }
+
+        default: {
             throw new Error(`${contractName} is not supported.`);
+        }
     }
 
     printInfo(`${contractName} deployed`, JSON.stringify(contractConfig, null, 2));
 }
 
-async function mainProcessor(contractName, options, processor) {
+async function upgrade(contractName, policy, config, chain, options) {
+    const [keypair, client] = getWallet(chain, options);
+    const { packageDependencies } = options;
+    options.policy = policy;
+
+    await printWalletInfo(keypair, client, chain, options);
```

Comment:
duplication, do this in the mainProcessor, and pass in the keypair to deploy/upgrade